File: backend/app/utils/iban.py
import re

# Expected IBAN lengths per country code (ISO 13616)
IBAN_LENGTHS: dict[str, int] = {
    "DE": 22,
    "AT": 20,
    "CH": 21,
    "NL": 18,
    "FR": 27,
    "BE": 16,
    "LU": 20,
    "IT": 27,
    "ES": 24,
}


def validate_iban(iban: str) -> tuple[bool, str]:
    """Validate an IBAN. Returns (True, cleaned_iban) or (False, error_message)."""
    # Clean: remove whitespace, dashes, uppercase
    cleaned = re.sub(r"[\s\-]", "", iban).upper()

    if not cleaned:
        return False, "IBAN darf nicht leer sein"

    # Basic format: 2 letters + 2 digits + alphanumeric
    if not re.match(r"^[A-Z]{2}\d{2}[A-Z0-9]+$", cleaned):
        return False, "IBAN hat ein ungueltiges Format"

    country = cleaned[:2]

    # Country-specific length check
    expected_len = IBAN_LENGTHS.get(country)
    if expected_len is None:
        return False, f"Laendercode '{country}' wird nicht unterstuetzt"

    if len(cleaned) != expected_len:
        return (
            False,
            f"IBAN fuer {country} muss {expected_len} Zeichen haben, hat aber {len(cleaned)}",
        )

    # Modulo 97 check (ISO 13616)
    rearranged = cleaned[4:] + cleaned[:4]
    numeric = ""
    for char in rearranged:
        if char.isdigit():
            numeric += char
        else:
            numeric += str(ord(char) - 55)

    if int(numeric) % 97 != 1:
        return False, "IBAN-Pruefsumme ist ungueltig"

    return True, cleaned


def format_iban(iban: str) -> str:
    """Format IBAN into groups of 4: 'DE89 3704 0044 0532 0130 00'."""
    cleaned = re.sub(r"[\s\-]", "", iban).upper()
    return " ".join(cleaned[i : i + 4] for i in range(0, len(cleaned), 4))


def mask_iban(iban: str) -> str:
    """Mask IBAN: show first 4 + last 4, rest as stars in 4-groups.
    'DE89 **** **** **** **30 00'
    """
    cleaned = re.sub(r"[\s\-]", "", iban).upper()
    if len(cleaned) <= 8:
        return format_iban(cleaned)

    first4 = cleaned[:4]
    last4 = cleaned[-4:]
    middle_len = len(cleaned) - 8
    masked = first4 + "*" * middle_len + last4

    return " ".join(masked[i : i + 4] for i in range(0, len(masked), 4))

File: php-ionos/app/iban.php
<?php
/**
 * IBAN-Validierung (ISO 13616, Modulo 97), Formatierung und Maskierung.
 * Portiert aus backend/app/utils/iban.py.
 */

declare(strict_types=1);

const IBAN_LENGTHS = [
    'DE' => 22, 'AT' => 20, 'CH' => 21, 'NL' => 18, 'FR' => 27,
    'BE' => 16, 'LU' => 20, 'IT' => 27, 'ES' => 24,
];

/** Gibt [true, bereinigte IBAN] oder [false, Fehlermeldung] zurück. */
function validate_iban(string $iban): array
{
    $cleaned = strtoupper(preg_replace('/[\s\-]/', '', $iban));

    if ($cleaned === '') {
        return [false, 'IBAN darf nicht leer sein'];
    }
    if (!preg_match('/^[A-Z]{2}\d{2}[A-Z0-9]+$/', $cleaned)) {
        return [false, 'IBAN hat ein ungültiges Format'];
    }

    $country = substr($cleaned, 0, 2);
    if (!isset(IBAN_LENGTHS[$country])) {
        return [false, "Ländercode '$country' wird nicht unterstützt"];
    }
    $expected = IBAN_LENGTHS[$country];
    if (strlen($cleaned) !== $expected) {
        return [false, "IBAN für $country muss $expected Zeichen haben, hat aber " . strlen($cleaned)];
    }

    // Modulo-97-Prüfung mit bcmod-freier Implementierung (chunkweise)
    $rearranged = substr($cleaned, 4) . substr($cleaned, 0, 4);
    $numeric = '';
    foreach (str_split($rearranged) as $char) {
        $numeric .= ctype_digit($char) ? $char : (string)(ord($char) - 55);
    }

    $remainder = 0;
    foreach (str_split($numeric, 7) as $chunk) {
        $remainder = (int)(($remainder . $chunk)) % 97;
    }
    if ($remainder !== 1) {
        return [false, 'IBAN-Prüfsumme ist ungültig'];
    }

    return [true, $cleaned];
}

/** 'DE89 3704 0044 0532 0130 00' */
function format_iban(string $iban): string
{
    $cleaned = strtoupper(preg_replace('/[\s\-]/', '', $iban));
    return implode(' ', str_split($cleaned, 4));
}

/** Erste 4 + letzte 4 Zeichen sichtbar, Rest maskiert. */
function mask_iban(string $iban): string
{
    $cleaned = strtoupper(preg_replace('/[\s\-]/', '', $iban));
    if (strlen($cleaned) <= 8) {
        return format_iban($cleaned);
    }
    $masked = substr($cleaned, 0, 4)
        . str_repeat('*', strlen($cleaned) - 8)
        . substr($cleaned, -4);
    return implode(' ', str_split($masked, 4));
}
